core-bundle/src/EventListener/DoctrineSchema/AdjustSearchUrlLengthListener.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener\DoctrineSchema;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Schema\SchemaException;
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\ORM\Tools\Event\GenerateSchemaEventArgs;
+
+/**
+ * Adjusts the length of tl_search.url if innodb_large_prefix is not enabled (#4615).
+ *
+ * @internal
+ */
+class AdjustSearchUrlLengthListener
+{
+    private $connection;
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    public function __invoke(GenerateSchemaEventArgs $event): void
+    {
+        // Get the tl_search table definition
+        try {
+            $table = $event->getSchema()->getTable('tl_search');
+        } catch (SchemaException $e) {
+            if (SchemaException::TABLE_DOESNT_EXIST === $e->getCode()) {
+                return;
+            }
+
+            throw $e;
+        }
+
+        // Get the tl_search.url column definition
+        try {
+            $column = $table->getColumn('url');
+        } catch (SchemaException $e) {
+            if (SchemaException::COLUMN_DOESNT_EXIST === $e->getCode()) {
+                return;
+            }
+
+            throw $e;
+        }
+
+        // Check if the field has an index
+        try {
+            $table->getIndex('url');
+        } catch (SchemaException $e) {
+            if (SchemaException::INDEX_DOESNT_EXIST === $e->getCode()) {
+                return;
+            }
+
+            throw $e;
+        }
+
+        // Get maximum index size for this table
+        $maxIndexSize = $this->getMaximumIndexSize($table);
+
+        // Reduce maximum length if collation is not "ascii_bin"
+        if ('ascii_bin' !== $column->getPlatformOption('collation')) {
+            $bytesPerChar = 'utf8mb4' === $table->getOption('charset') ? 4 : 3;
+            $maxIndexSize = floor($maxIndexSize / $bytesPerChar);
+        }
+
+        if ($column->getLength() <= $maxIndexSize) {
+            return;
+        }
+
+        // Set the length
+        $column->setLength($maxIndexSize);
```

Comment:
Should we trigger a warning here?

```suggestion
        $column->setLength($maxIndexSize);

        trigger_error(sprintf('The length of the tl_search.url column was reduced to %s because your database does not support longer UNIQUE indexes.'), E_USER_WARNING);
```

Reply:
Not sure, we don't do that for the `tl_files` index either currently.

Reply:
That’s a different story IMO because we only reduce the size on the (non-unique-)index there (which should not change the behavior).

Reply:
True. But the warning makes the install tool unusable in debug mode.

Reply:
How about a log entry instead at least?

Reply:
Would be fine for me too.
A deprecation might also work?

Reply:
> A deprecation might also work?

So increasing the minimum requirements for Contao 5.0? Would make sense to me too.

Reply:
I have added the deprecation now. However, the message is a bit clunky and I am also not sure about the tests. In order to accept deprecations I had to add `@group legacy`, which seems wrong.

Reply:
It is fine IMO and we have many `@group legacy` tests that are not fully legacy.